File: things/shadow/shadow.go
// Copyright 2021 Scott Feldman (sfeldma@gmail.com). All rights reserved.
// Use of this source code is governed by a BSD-style license that can be found
// in the LICENSE file.

// Shadow is a Thing's shadow.
package shadow

import (
	"github.com/scottfeldman/merle"
	"html/template"
	"net/http"
)

var templ *template.Template

func init() {
	templ = template.Must(template.ParseFiles("web/templates/shadow.html"))
}

type shadow struct {
	merle.Thing
}

func (s *shadow) init() error {
	return s.ListenForThings(1, "[*][*][*]")
}

func (s *shadow) run() {
	for {}
}

func (s *shadow) home(w http.ResponseWriter, r *http.Request) {
	templ.Execute(w, s.HomeParams(r))
}

func NewThing(id, model, name string) *merle.Thing {
	s := &shadow{}

	s.Init = s.init
	s.Run = s.run
	s.Home = s.home

	return s.InitThing(id, model, name)
}

File: web/js/shadow.js
// Copyright 2021 Scott Feldman (sfeldma@gmail.com). All rights reserved.
// Use of this source code is governed by a BSD-style license that can be found
// in the LICENSE file.

function sendForThings() {
	conn.send(JSON.stringify({Msg: "GetThings"}))
}

function show(id) {
	var iframe = document.getElementById("thing")
	iframe.src = "/" + encodeURIComponent(id)
}

function getThing(msg) {
	if (msg.Things != null) {
		thing = msg.Things[0]
		show(thing.Id)
	}
}

function Run(scheme, host, id) {

	conn = new WebSocket(scheme + host + "/ws/" + id)

	conn.onopen = function(evt) {
		sendForThings()
	}

	conn.onclose = function(evt) {
		location.reload(true)
	}

	conn.onmessage = function(evt) {
		var msg = JSON.parse(evt.data)

		console.log('event', msg)

		switch(msg.Msg) {
		case "ReplyThings":
			getThing(msg)
			break
		case "SpamStatus":
			show(msg.Id)
			break
		}
	}
}
